### ISSUE TITLE
Stack Items should be popped explicitly, not by dropping

### ISSUE THREAD

**Issue by sanboxrunner** [Gekkio/imgui-rs#257]
Hello!

Right now, to make many style choices, one does the following:

```rs
                    let _ = ui.push_item_width(8.0);
                    ui.same_line_with_spacing(ui.get_item_rect_size()[0], 12.0);
                    ui.slider_float(
                        im_str!("Grid Thickness"),
                        &mut gameplay.game_colors.grid_line_width,
                        0.0,
                        0.5,
                    )
                    .build();
                    // drop(item_width);
```

Commenting or uncommenting that `drop` does nothing in this case, of course, but if there was code after it that we wanted to do things with, we'd want to "drop" it there.

tl;dr: this is a bad API design and we should be pushing/popping ourselves.

The idea of doing it automagically makes sense, but unfortunately, it also doesn't. To call `drop`, is actually calling `pop` on that stack. This leads to a few problems.

We have to have `must_use` on for the return value, which is a sign that we're *already* dependent on the user for our API to work. This is because, of course, we are. If the user didn't mark that return value to be dropped, we'll pop it right out immediately and their efforts will be wasted. Therefore, we've told them that they `must_use` it. 

This has a few problems: If they're scoping in a way that our API works nicely with, by having all these style vars inside `{}`, so drop gets called implicitly, they'll need to name their variable with a `_` at the beginning. Not *just* a `_` however! As many Rustaceans *don't* know, since it doesn't come up much, when you do `_`, the value is immediately dropped and isn't actually assigned (saving stack space). This means the *least* intrusive way to use our API is to name a return for a throwaway value something like `_hey`. This makes code look ugly, in my opinion. 

If users aren't scoping these, and are calling drop themselves, then we've got another problem. First, using `drop` to change state exterior to the object being dropped is, in my opinion, not very good. The user has to remember what `drop` is doing, which, since there's no documentation on the type itself, requires some digging around to figure out. Moreover, and I think this is the final nail in the coffin, to me, we're calling `drop` on a specific Item, as if we are doing something to that item. This could easily lead a user to not understand that they're dealing with an internal Stack, and that the item itself actually has no intelligence on it whatsoever.

Finally, therefore, I propose we remove these drops and instead make the `push` versions of these functions available, wrapped in an appropriate `Option` or `Result` return type. If the user messes it up, well, let them!

If you find that unacceptable, we could easily just add a `Pop` method which *takes* a returned StyleToken (and all the other tokens of this sort), and then calls the Pop on it. We can then turn the `drop` implementation into a warning, perhaps, if the user has correctly balanced their pushes/pops.

I am more than glad to do this work myself -- it seems relatively trivial.

Thank you!

**Comment by Gekkio** [comment]
Great minds think alike :wink:

https://github.com/Gekkio/imgui-rs/commit/46d099f40f01c191a4b8212f0e371d77f7dd18b5

0.2 already requires explicit pop/end calls, and `Drop` simply panics by default if the corresponding call hasn't been made.
Panicking in Drop is normally a bit of a no-no, but in this case dear imgui would terminate the process without a stack trace in all the scenarios anyway. By panicking "early" with Drop, we get a stack trace with some hints about where to start looking.

**Comment by sanboxrunner** [comment]
Sounds *perfect* to me. As long as we make it clear where it's coming from, that's great.

Love you work, very happy to see you were going in this direction already! I've been using this tool fairly substantially -- if you mark any work as being something you could use help with, I'd more than gladly contribute! (Been using imgui in...4 different languages now? I quite like it :) )

**Comment by sanboxrunner** [comment]
Oh I can close as you're working on it.

**Comment by Gekkio** [comment]
Cool :smile: If I remember correctly, all push/pop and begin/end pairs have already been updated in 0.2, so everything should work fine in the master branch.

Here's also some more information about the topic, including examples of what happens if you forget to pop:
https://github.com/Gekkio/imgui-rs/issues/238#issuecomment-511326759
